### PR TITLE
Raw service

### PR DESCRIPTION
**Pull request by bmarty**
Create a RawService to allow clients of the SDK to retrieve plain text from any URL, with a cache management.

It will allow clients to retrieved data that are not part of the Matrix specification, especially from the Well-known file.

The Database migration removed field that has been added to HomeServerCapability and a a RawCache table to store the fetched data.

### CODE REVIEW

**Comment by bmarty** [comment]
I'm working to extract this service from the session

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/DefaultRawService.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                    callback = matrixCallback
+                }
+                .executeBy(taskExecutor)
+    }
+
+    override fun getWellknown(userId: String,
+                              matrixCallback: MatrixCallback<String>): Cancelable {
+        val homeServerDomain = userId.substringAfter(":")
+        return getUrl(
+                "https://$homeServerDomain/.well-known/matrix/client",
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/DefaultRawService.kt:47]: Constant?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/DefaultGetUrlTask.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            apiCall = rawAPI.getUrl(url)
+        }
+                .string()
+    }
+
+    private suspend fun doRequestWithCache(url: String, validityDurationInMillis: Long, strict: Boolean): String {
+        // Get data from cache
+        var dataFromCache: String? = null
+        var isCacheValid = false
+        monarchy.awaitTransaction { realm ->
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/DefaultGetUrlTask.kt:72]: why using awaitTransaction here?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/RawModule.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        @GlobalDatabase
+        fun providesMonarchy(@GlobalDatabase realmConfiguration: RealmConfiguration): Monarchy {
+            return Monarchy.Builder()
+                    .setRealmConfiguration(realmConfiguration)
+                    .build()
+        }
+
+        @JvmStatic
+        @Provides
+        @GlobalDatabase
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/RawModule.kt:52]: Add MatrixScope I guess

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/RawModule.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+@Module
+internal abstract class RawModule {
+
+    @Module
+    companion object {
+        private const val DB_ALIAS = "matrix-sdk-global"
+
+        @JvmStatic
+        @Provides
+        @GlobalDatabase
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/RawModule.kt:43]: Add MatrixScope I guess

File: vector/src/main/java/im/vector/app/features/home/room/detail/RoomDetailViewModel.kt [hunk 5]
```diff
@@ -349,7 +352,13 @@ class RoomDetailViewModel @AssistedInject constructor(
             val roomId: String = room.roomId
             val confId = roomId.substring(1, roomId.indexOf(":") - 1) + widgetSessionId.toLowerCase(VectorLocale.applicationLocale)
 
-            val jitsiDomain = session.getHomeServerCapabilities().preferredJitsiDomain ?: stringProvider.getString(R.string.preferred_jitsi_domain)
+            val preferredJitsiDomain = tryThis {
+                awaitCallback<String> { rawService.getWellknown(session.myUserId, it) }
+                        .let { ElementWellKnownMapper.from(it) }
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/features/home/room/detail/RoomDetailViewModel.kt:357]: This is used in multiple spaces, maybe we should provide an  extension method?

File: vector/src/main/java/im/vector/app/features/homeserver/WellKnownExt.kt [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        @Json(name = "io.element.e2ee")
+        val elementE2E: E2EWellKnownConfig? = null,
+
+        @Json(name = "im.vector.riot.e2ee")
+        val riotE2E: E2EWellKnownConfig? = null
+)
+
+object ElementWellKnownMapper {
+
+    val adapter: JsonAdapter<ElementWellKnown> = MoshiProvider.providesMoshi().adapter(ElementWellKnown::class.java)
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/features/homeserver/WellKnownExt.kt:45]: We are exposing Moshi from the sdk, not sure this is really great

**Review by ganfra — COMMENTED** [review]
Some remarks, otherwise good work

File: vector/src/main/java/im/vector/app/features/homeserver/WellKnownExt.kt (hunk 6 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/homeserver/WellKnownExt.kt:45]: Yes, I know, but it is already the case elsewhere... We may stop doing that later

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/DefaultRawService.kt (hunk 1 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/DefaultRawService.kt:47]: We can change that later

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/DefaultGetUrlTask.kt (hunk 2 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/DefaultGetUrlTask.kt:72]: my bad

File: vector/src/main/java/im/vector/app/features/home/room/detail/RoomDetailViewModel.kt (hunk 5 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/home/room/detail/RoomDetailViewModel.kt]: Done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/RawModule.kt (hunk 4 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/RawModule.kt:43]: Later

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/RawModule.kt (hunk 3 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/raw/RawModule.kt:52]: Later

**Review by ganfra — APPROVED** [review]
LGTM thanks!

**Pull request merged**